pkg/agent/memberlist/cluster_test.go

```diff
@@ -0,0 +1,741 @@
+//cluster_test.go:612: Node: node2, egressNum: 317
+//cluster_test.go:612: Node: node3, egressNum: 335
+//=== RUN   TestCluster_ShouldSelectByConsistentHash/select_node_from_alive_nodes-10000-Egresses
+//=== CONT  TestCluster_ShouldSelectByConsistentHash
+//cluster_test.go:612: Node: node1, egressNum: 3318
+//cluster_test.go:612: Node: node2, egressNum: 3469
+//cluster_test.go:612: Node: node3, egressNum: 3213
+//--- PASS: TestCluster_ShouldSelectByConsistentHash (0.10s)
+
+// replicas = 50
+//=== RUN   TestCluster_ShouldSelectByConsistentHash/select_node_from_alive_nodes-3-Egresses
+//=== CONT  TestCluster_ShouldSelectByConsistentHash
+//cluster_test.go:612: Node: node1, egressNum: 1
+//cluster_test.go:612: Node: node2, egressNum: 2
+//cluster_test.go:612: Node: node3, egressNum: 0
+//=== RUN   TestCluster_ShouldSelectByConsistentHash/select_node_from_alive_nodes-10-Egresses
+//=== CONT  TestCluster_ShouldSelectByConsistentHash
+//cluster_test.go:612: Node: node1, egressNum: 1
+//cluster_test.go:612: Node: node2, egressNum: 4
+//cluster_test.go:612: Node: node3, egressNum: 5
+//=== RUN   TestCluster_ShouldSelectByConsistentHash/select_node_from_alive_nodes-100-Egresses
+//=== CONT  TestCluster_ShouldSelectByConsistentHash
+//cluster_test.go:612: Node: node1, egressNum: 30
+//cluster_test.go:612: Node: node2, egressNum: 32
+//cluster_test.go:612: Node: node3, egressNum: 38
+//=== RUN   TestCluster_ShouldSelectByConsistentHash/select_node_from_alive_nodes-1000-Egresses
+//=== CONT  TestCluster_ShouldSelectByConsistentHash
+//cluster_test.go:612: Node: node1, egressNum: 345
+//cluster_test.go:612: Node: node2, egressNum: 297
+//cluster_test.go:612: Node: node3, egressNum: 358
+//=== RUN   TestCluster_ShouldSelectByConsistentHash/select_node_from_alive_nodes-10000-Egresses
+//=== CONT  TestCluster_ShouldSelectByConsistentHash
+//cluster_test.go:612: Node: node1, egressNum: 3437
+//cluster_test.go:612: Node: node2, egressNum: 2930
+//cluster_test.go:612: Node: node3, egressNum: 3633
+//--- PASS: TestCluster_ShouldSelectByConsistentHash (0.10s)
+//https://github.com/golang/groupcache/issues/29
+func TestCluster_ShouldSelectEgress(t *testing.T) {
+	nodes := []string{"node1", "node2", "node3"}
+	consistentHash := consistenthash.New(defaultVirtualNodeReplicas, nil)
+	consistentHash.Add(nodes...)
+
+	nodeSelectedForEgress := func(consistentHash *consistenthash.Map, egressName, nodeName string) bool {
+		return consistentHash.Get(egressName) == nodeName
+	}
+
+	checkNum := func(count int, localNode string) int {
+		totalNum := 0
+		for i := 0; i < count; i++ {
+			egressName := fmt.Sprintf("egress-%d", i)
+			if nodeSelectedForEgress(consistentHash, egressName, localNode) {
+				totalNum++
+			}
+		}
+		return totalNum
+	}
+
+	checkSum := func(egressNum int) int {
+		count := 0
+		for _, node := range nodes {
+			num := checkNum(egressNum, node)
+			count += num
+			t.Logf("Node: %s, egressNum: %d", node, num)
+		}
+		return count
+	}
```

Comment:
I think the current test just ensures `consistenthash.Map` works as expected, not covering any logic in antrea. It didn't even call any Antrea function.
To provide real value, I think the test can test whether `ShouldSelectEgress` can get expected result with given input. For example:
```
	testCases := []struct {
		name      string
		nodeNum   int
		egressIP    string
		expectedNode string
	}{
		{
			name: "none Node",
			nodeNum: 0,
			egressIP: "1.1.1.1",
			expectedNode: "",
		},
		{
			name: "single Node",
			nodeNum: 1,
			egressIP: "1.1.1.1",
			expectedNode: "node1",
		},
		{
			name: "3 Nodes",
			nodeNum: 3,
			egressIP: "1.1.1.1",
			expectedNode: "nodeX",
		},
		{
			name: "10 Nodes",
			nodeNum: 10,
			egressIP: "1.1.1.1",
			expectedNode: "nodeX",
		},
	}
```
It ensures none Node will be selected when there is no eligable Node, the only Node will be selected when there is single eligable Node, and the same Node will be selected when there are multiple Nodes. Especially the latter two, which can guarantee we can get consistent result across antrea versions. Otherwise an accidental change to the implementation of ShouldSelectNode will cause agents can inconsistent result while upgrading, which causes datapath disruption.

Reply:
done